src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CSupport.java

```diff
@@ -147,12 +157,30 @@ public static ForeignLinker getSystemLinker() {
         /**
          * Copies this C {@code va_list}.
          *
+         * @apiNote that this method only copies the va list 'view' and not any argument space it may manage.
+         * That means that if this va list was created with the {@link #make(Consumer)} method, closing
+         * this va list will also free the argument space, and make the copy unusable.
```

Comment:
A VaList is really only a 'view' of a set of arguments stored elsewhere. In native code the place where the arguments are stored (i.e. the 'argument space') is on the stack, so the memory will get cleaned up automatically. But, if we create a VaList on the Java side we need to do an explicit off-heap allocation and later call free to clean up the memory that stores the arguments. This argument space is freed when calling `close()` on the VaList.

So, if you `copy()` a VaList, only the 'view' part is copied, but not the memory that holds the arguments.

Reply:
I'm thinking of maybe dropping the VaList::make overload that doesn't take a NativeScope, so that the memory for the arguments will always be managed by a NativeScope, and it doesn't have to piggyback on VaList::close

Reply:
I'll try and add a test for this as well